File: server2/server2.go
package main

import (
	"fmt"
	"graphql-mock/stuff"
	"io/ioutil"
	"math/rand"

	"github.com/gin-gonic/gin"
	graphql "github.com/graph-gophers/graphql-go"
	"github.com/graph-gophers/graphql-go/relay"
)

// represents a die
type randomDie struct{ numSides int32 }

// constructor
func newRandomDie(numSides int32) *randomDie {
	rd := randomDie{numSides: numSides}
	return &rd
}

func (r *randomDie) NumSides() int32 {
	return r.numSides
}

func (r *randomDie) RollOnce() int32 {
	return 1 + rand.Int31n(r.numSides)
}

func (r *randomDie) Roll(args struct{ NumRolls int32 }) []int32 {
	output := make([]int32, 0, args.NumRolls)
	for i := int32(0); i < args.NumRolls; i++ {
		output = append(output, r.RollOnce())
	}
	return output
}

type root struct{}

func (*root) GetDie(args struct{ NumSides int32 }) *randomDie {
	return newRandomDie(args.NumSides)
}

// Construct a schema, using GraphQL schema language
var schemaBytes, _ = ioutil.ReadFile("server2.graphql")
var schema = graphql.MustParseSchema(string(schemaBytes), &root{}, graphql.UseStringDescriptions())

// GIN handler
func gopherGraphQLHandler() gin.HandlerFunc {
	r := &relay.Handler{Schema: schema}
	return func(c *gin.Context) {
		r.ServeHTTP(c.Writer, c.Request)
	}
}

func main() {
	// r := gin.New()
	r := gin.Default()
	r.POST("/graphql", gopherGraphQLHandler())
	r.GET("/", stuff.PlaygroundHandler("GraphQL playground", "/graphql"))

	fmt.Println(`
	Running a GraphQL playground http://localhost:4000/
	GraphQL end point            http://localhost:4000/graphql
	`)

	r.Run(":4000")
}

File: server2/server2.js
// Running GraphQL API server
// npm install express express-graphql graphql cors
// https://graphql.org/graphql-js/object-types/

var fs = require('fs')
var express = require('express')
var { graphqlHTTP } = require('express-graphql')
var { buildSchema } = require('graphql')
var cors = require('cors')


// Represents a die
class RandomDie {
    constructor (numSides) {
        this.numSides = numSides
    }

    rollOnce () {
        return 1 + Math.floor(Math.random() * this.numSides)
    }

    roll ({ numRolls }) {
        var output = []
        for (var i = 0; i < numRolls; i++) {
            output.push(this.rollOnce())
        }
        return output
    }
}

// The root provides the top-level API endpoints
var root = {
    getDie: ({ numSides } ) => {
       return new RandomDie(numSides)
    },
}

// Construct a schema, using GraphQL schema language
var schema = buildSchema(fs.readFileSync('server2.graphql', 'utf8'))

var app = express()
app.use(cors())
app.use( '/graphql', graphqlHTTP({
        schema: schema,
        rootValue: root,
        graphiql: true
    })
)
console.log('Running a GraphQL API server at http://localhost:4000/graphql')
app.listen(4000)
